lib/kubernetes-deploy/kubernetes_resource/custom_resource_definition.rb

```diff
@@ -2,8 +2,86 @@
 module KubernetesDeploy
   class CustomResourceDefinition < KubernetesResource
     TIMEOUT = 2.minutes
+    ROLLOUT_CONFIG_ANNOTATION = "kubernetes-deploy.shopify.io/monitor-instance-rollout"
     GLOBAL = true
 
+    class RolloutConfig
+      attr_reader :success_queries, :failure_queries
+
+      def initialize(rollout_config)
+        config_json = JSON.parse(rollout_config)
+        config = {
+          success_queries: config_json["success_queries"] || default_success_query,
+          failure_queries: config_json["failure_queries"] || default_failure_query,
+        }.deep_symbolize_keys
+
+        # Preemptively create JsonPath objects
+        config[:success_queries].map! do |query|
+          query.update(query) { |k, v| k == :path ? JsonPath.new(v) : v }
+        end
+        config[:failure_queries].map! do |query|
+          query.update(query) { |k, v| k == :path || k == :error_msg_path ? JsonPath.new(v) : v }
+        end
+
+        validate_config(config)
+        @success_queries = config[:success_queries]
+        @failure_queries = config[:failure_queries]
+      rescue JSON::ParserError
+        raise FatalDeploymentError, "custom rollout params are not valid JSON: '#{rollout_config}'"
+      rescue RuntimeError => e
+        raise FatalDeploymentError, "error creating jsonpath objects, failed with: #{e}"
+      end
+
+      def deploy_succeeded?(instance_data)
+        @success_queries.all? do |query|
+          query[:path].on(instance_data).first == query[:value]
```

Comment:
Why `.first`? When might we get multiple matches?

Reply:
This is a peculiarity of `jsonpath`

Reply:
I guess when we will get multiple matches depends entirely on the path they give us and the structure it is matching... it's very easy to build queries that will match multiple results given a multi-level object, and [the gem's docs](https://github.com/joshbuddy/jsonpath) have some examples. If we did get multiple results, do you think the correct behaviour be `.first` or `.all?`?
 I also noticed in the docs that you can do `query.first(instance_data)` directly if that's what we want.

Reply:
I'm imagining that most jsonpaths will be fully specified (from root of document to target field) so am not too concerned about dealing with multiple matches. I think that would cause too much confusion in the implementation without a corresponding increase in usability; users should be able to get by with only single field matches